### DISCUSSION TITLE
multi_read_vars() reading from incorrect address

### DISCUSSION THREAD

**Discussion by Rahul14189** [gijzelaerr/python-snap7#457] [Q&A] [answered]
Hi, I am trying to read multiple address using multi_read_vars(), db_read() is reading the address correctly, but multi_read_vars() is giving undesired value, and sometime says address does not exists,  i have followed the example from multi_read.py,   This is my request parameter:
       data_items[i].Area = ctypes.c_int32(S7AreaDB)
        data_items[i].WordLen = ctypes.c_int32(S7WLWord)
        data_items[i].Result = ctypes.c_int32(0)
        data_items[i].DBNumber = ctypes.c_int32(2)
        data_items[i].Start = ctypes.c_int32(2)
        data_items[i].Amount = ctypes.c_int32(4)

    for di in data_items:
        # create the buffer
        buffer = ctypes.create_string_buffer(di.Amount)

        # cast the pointer to the buffer to the required type
        pBuffer = ctypes.cast(ctypes.pointer(buffer),
                              ctypes.POINTER(ctypes.c_uint8))
        di.pData = pBuffer

result, data_items_res = plc.read_multi_vars(data_items)

        for di in data_items_res:
            check_error(di.Result)

byte_to_value = [util.get_real, util.get_int, util.get_bool, util.get_dword]

for i in range(0, len(data_items_res)):
            btv = byte_to_value[i]
            di = data_items_res[i]
            value = btv(di.pData, 0)
            result.append(value)
        print(result_values)

AND my request parameter for db_read(2, 2, 4),      My issue is, i am sending correct request parameter which works on db_read(), but it is not working for multi_read_vars() , so is it, that i am missing something here !!

**Comment by nikteliy** [comment]
Hi @Rahul14189 
Could you please share the full code of request and properly format it, so we can better understand what is causing the error?

**Reply by Rahul14189** [reply]
Hi, I have attached the whole code below:

def find_datatype(memory_address):
    if memory_address.split(',')[1][0] == 'X':
        return S7WLBit
    elif memory_address.split(',')[1][0] == 'D':
        return S7WLWord

if __name__ == '__main__':

    client = snap7.client.Client()
    client.connect('10.100.5.2', 0, 2)

    if client.get_connected():

        tag_addresses = ["DB1,X6.0", "DB1,D2"]

        data_items = (S7DataItem * len(tag_addresses))()

        for i in range(len(data_items)):
            data_items[i].Area = ctypes.c_int32(S7AreaDB)
            data_items[i].WordLen = ctypes.c_int32(find_datatype(tag_addresses[i]))
            data_items[i].Result = ctypes.c_int32(0)
            data_items[i].DBNumber = ctypes.c_int32(int(tag_addresses[i].split(',')[0].split('DB')[-1]))
            data_items[i].Start = ctypes.c_int32(int(tag_addresses[i].split(',')[1].split('.')[0][1]))
            data_items[i].Amount = ctypes.c_int32(1 if tag_addresses[i].startswith("DB1,X") else 4)

        for di in data_items:
            buffer = ctypes.create_string_buffer(di.Amount)

            pBuffer = ctypes.cast(ctypes.pointer(buffer),
                                  ctypes.POINTER(ctypes.c_uint8))
            di.pData = pBuffer

        result, data_items_res = client.read_multi_vars(data_items)

        for di in data_items_res:
            check_error(di.Result)

        result_values = {}

        byte_to_value = [util.get_bool, util.get_dword]

        for i in range(0, len(data_items_res)):
            btv = byte_to_value[i]
            di = data_items_res[i]
            value = btv(di.pData, 0)
            result_values[tag_addresses[i]] = value
        print(result_values)
        client.disconnect()
        client.destroy()

**Comment by nikteliy** [comment, marked as answer]
I've tested your code on my PLC, and it works. You can see it on the screenshot:
![img1](https://github.com/gijzelaerr/python-snap7/assets/52915342/293b748d-97c2-47db-b168-b078ddaad030)

Here's the code that I've run:

```

def find_datatype(memory_address):
    if memory_address.split(',')[1][0] == 'X':
        return S7WLBit
    elif memory_address.split(',')[1][0] == 'D':
        return S7WLWord

if __name__ == '__main__':

    client = snap7.client.Client()
    client.connect('10.10.10.100', 0, 1)

    if client.get_connected():

        tag_addresses = ["DB1,X6.0", "DB1,D2"]

        data_items = (S7DataItem * len(tag_addresses))()

        for i in range(len(data_items)):
            data_items[i].Area = ctypes.c_int32(S7AreaDB)
            data_items[i].WordLen = ctypes.c_int32(find_datatype(tag_addresses[i]))
            data_items[i].Result = ctypes.c_int32(0)
            data_items[i].DBNumber = ctypes.c_int32(int(tag_addresses[i].split(',')[0].split('DB')[-1]))
            data_items[i].Start = ctypes.c_int32(int(tag_addresses[i].split(',')[1].split('.')[0][1]))
            data_items[i].Amount = ctypes.c_int32(1 if tag_addresses[i].startswith("DB1,X") else 4)

        for di in data_items:
            buffer = ctypes.create_string_buffer(di.Amount)

            pBuffer = ctypes.cast(ctypes.pointer(buffer),
                                ctypes.POINTER(ctypes.c_uint8))
            di.pData = pBuffer

        result, data_items_res = client.read_multi_vars(data_items)

        for di in data_items_res:
            check_error(di.Result)

        result_values = {}

        byte_to_value = [util.get_bool, util.get_dword]

        for i in range(0, len(data_items_res)):
            btv = byte_to_value[i]
            di = data_items_res[i]
            if i == 0 :
                value = btv(di.pData, 0, 0)
            else:
                value = btv(di.pData, 0)
            result_values[tag_addresses[i]] = value
        
        print(f'read_multi: {result_values}')
        print(f'db_read: {util.get_dword(client.db_read(1, 2, 4), 0)}')

        client.disconnect()
        client.destroy()
```

**Reply by Rahul14189** [reply]
Thanks, it worked now, there was issue at PLC end, and at my end "byte_to_value" order was not correct as i was using [util.get_real,  util.get_dword, util.get_bool] in my other script,

However i see that, at once we cannot send more than 20 tag_address, 
\script\lib\site-packages\snap7\common.py", line 89, in check_error
    raise RuntimeError(error)
RuntimeError: b'CLI : too may items (>20) in multi read/write'
